### DISCUSSION TITLE
Expose parent reference in actor methods

### DISCUSSION THREAD

**Discussion by parker-codes** [statelyai/xstate#2715] [Ideas]
I'm reaching a pain point where my machines need access to others' context. In my case, a global app machine has a tree of actors as children. When a page machine (actor) needs to reference user data in another machine (actor), it would be great to reference it directly. Otherwise, it seems like I would have to broadcast and store overlaps of data in many machines.  In my case I'd use a reference to the parent and traverse over to a sibling, but the developer-defined network of actors is the key here. As long as they link pieces properly they can access data. I would be worried about breaking things if I reorganized where actors are located, but this type of API probably doesn't need to be concerned with that. It's akin to a path system.

@mattpocock and I have discussed this a couple of times recently in Discord channels. 

Here's a proposed API:
```ts
assign((ctx, event, { parent }) => {
  console.log(parent.getSnapshot().context.someSibling.getContext().context);
})
```

Because performance and security are concerns when passing this reference, there could be a flag to enable:
```ts
spawn(someMachine, { withParent: true })
```

It seems like this would be best exposed in all action types, services, and guards.

**Comment by parker-codes** [comment]
Tidbits from conversation with @davidkpiano:

David:
> Well, you wouldn't want to give access to the parent by default; becomes a typing nightmare and is generally unsafe
What we probably should do instead is this:
> ```ts
> assign((ctx, event, { self }) => {
>   return {
>     ref: spawn(createSomeMachine(/* parent = */ self), 'name')
>   }
> })
> ```
> Which is much less magical, but about as simple
> ```ts
> const createSomeMachine = (parent: YouCanStronglyTypeThisActorRef) => createMachine({
>   context: { parent },
>   // ...
> })
> ```

> I guess it's not a big issue
Since to execute an action, you need to provide it:
`action(theContext, theEvent, allTheMetadata /* this is strongly typed */)`
So if you're missing self then TS will complain 👍
So this probably won't be that big of a change 🤔
But yeah, no `parent`; only `self`

**Reply by davidkpiano** [reply]
In v4, this is currently easy to implement for `invoke: { src: (ctx, e, { self }) => ... }` but not easy to do for actions, because of the way actions are implemented (this can be improved).

Still thinking about the best way to do this so that it's available in actions too.

**Reply by parker-codes** [reply]
Since actions seem to be more on the composition side, does it make more sense to expose it in another method like

```ts
someAction: expose(parent => assign((ctx, event) => {
  return {
    ref: spawn(createSomeMachine(parent), 'name')
  }
}))
```

Edit: I don't like the look of this, and I realized that you mean the implementation is the difficult part, not the design.

**Comment by parker-codes** [comment]
I got some feedback in Discord about how a good practice for hooking up actors can be to have a (top-level) invoke. This solved a race condition, which was nice, and also kind of highlights that the lifetime of the actor is paired with its parent.

This type of architecture wouldn't work with this idea of exposing `self`. I mean, self could be sent via a message, but that strategy is not fun. I'm doing it a lot when broadcasting websocket connections down a tree and it's quite messy.

Maybe invoked services just need a different API that is more declarative, like:
```ts
    invoke: {
      id: 'connection',
      src: connectionMachine,
      canAccessParent: true,
    },
```

Again, this is backwards and more like my original idea, but at least it has to be defined in the parent definition itself, so there's some level of protection. It also doesn't solve the actual implementation, but I just don't want to get blocked if I follow a practice that seems like it could clear up some of my code.

**Reply by davidkpiano** [reply]
In your example, machines always have "access" to the parent, at least for sending events to it. The problem with defining this implicit access for all children, besides it not being type-safe, is it's not obvious _how_ to access the parent. How do you access it from within a promise, or an observable, or a callback?

So, taking some inspiration from Akka, the two ways an actor can know about the parent are:

1. Be sent a reference to the parent from an event 
2. Be _born_ (created) with that knowledge

What I'm thinking of to support (2) is something like this:

```js
actions: assign((ctx, e, { self }) => {
  return { ref: spawn(createSomeActor(/* parent: */ self)) }
})
```

**Reply by parker-codes** [reply]
Yeah, and that totally makes sense! What I was trying to point out is that these invoked machines seem like they fit better than spawned ones stored in context (the lifetime thing I pointed out as well as how it solved a race condition), so it would make sense to also support this type of feature between them. In my opinion it seems like these machines tied together would have more of a need to access parent context than for many use cases of spawn, like when creating an array of actors.

**Comment by parker-codes** [comment]
Is there potentially a way to achieve this now, before the right API gets merged? 

I thought about sending a reference of the machine to itself, but couldn't get it to work:
![Screen Shot 2021-11-16 at 8 13 27 PM](https://user-images.githubusercontent.com/17652149/142127994-24ee2970-2d7b-4082-9235-1fb5b34b06a4.png)

**Reply by Andarist** [reply]
It should work - could u create a repro case for the problem?

**Reply by parker-codes** [reply]
It looks like it DID work! The problem was the self-referential type errors. I still haven't solved that - I'm just using `any` for now.

**Comment by parker-codes** [comment]
![Screen Shot 2021-11-22 at 4 12 48 PM](https://user-images.githubusercontent.com/17652149/142948777-2aeb2274-b68e-472b-99cb-614fb051a1d3.png)


Until the feature that exposes `self` makes it into XState, I have to find a workaround for accessing state from other actors as well as message passing between siblings (or else top-level actors get REALLY messy with forwarding so many messages. I'd like to be able to do what @davidkpiano#4536 stated in a video one time about how "you don't have to go through a mutual friend every time - once your friend helps you exchange phone numbers, you can message each other". The quote is in my own words, but the concept is nice. The only problem is, only parent actors have this reference to be able to pass around. 

Even if we do expose `self` though, it seems like there will be a ton of race conditions exposed because we don't want machine logic executing until the relationships are built. So if I can expose the app reference to `actor1` and `actor2` that seems fine, but what about linking `actor1` to `actor2.1` which is one level deeper? This isn't just for message passing, as I said before. `actor2.1` might also reference values from `actor1` instead of passing giant payloads in events or holding duplicate state. The race condition I'm talking about might happen if I start my actors on an `instantiating` state to wait for the connections to be made, but then other actors start sending events. Using a queue system for every actor seems like a bad pattern too. Some of them really are quite simple.

One thought I had was to instantiate all actors at once and then construct the tree from those pieces. This becomes a chicken-and-egg problem quickly though because we want to sometimes create a bi-directional link or have a circular communication strategy - really, users should be able to construct the communication how they'd like.

Going back to something I've said previously - it seems like invoking child actors as services is a good practice, but then these references are moot except for parent-to-child access. If a child could access the parent reference automatically (behind a flag when being invoked or spawned for security) then we wouldn't run into these race conditions of INIT events, etc.

My goal here is to get some feedback on how I might structure this in the short term, but then to continue the conversation for the right API long term.
